### ISSUE TITLE
Loaded messages appear for a Receiver chaotically in 1-1 chat

### ISSUE THREAD

**Issue by mariia-skrypnyk** [status-im/status-react#21460]
# Bug Report
Follow up of https://github.com/status-im/status-mobile/pull/21411

## Problem 
[comment]: # (A problem description)

When we, as the receiver, expect a batch of new messages, they often appear chaotically in our chat. The last message may show up first, followed by the initial messages, and only then the ones in the middle of the conversation.

### Reproduction

1. Login to app as a Receiver (a member of Status community)
2. Login to another device as a Sender (a member of Status community)
3. As a Receiver do a logout and close the app.
4. As a Sender wait till Receiver status become offline and send 20+ messages to 1-1 chat with a Receiver.
5. As a Receiver login and open chat with a Sender.

#### Actual behavior

Chaotic behavior starts from the 00:05 sec

https://github.com/user-attachments/assets/20f0d3e3-38c5-439c-9645-48f7ff693519

[Receiver_chaotic.zip](https://github.com/user-attachments/files/17434652/Receiver_chaotic.zip)
[Sender_chaotic.zip](https://github.com/user-attachments/files/17434658/Sender_chaotic.zip)

#### Expected behavior

Messages appear from the eldest to the newest logically

### Additional Information

- Status version: nightly 18/10
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

**Comment by ilmotta** [comment]
@mariia-skrypnyk We can close this issue because there's no guarantee of message ordering, as far as I understand. The behavior is like this by design, so it's probably the same behavior in the desktop app.

**Comment by mariia-skrypnyk** [comment]
Hey @ilmotta !

Thanks for your comment.
Are you talking regarding an implementation design?
As I am not sure I saw such logic in Figma or somewhere. 
Sorry but just wanted to be sure I know the source of the expected result.

**Comment by ilmotta** [comment]
Hey @mariia-skrypnyk, sorry I used the word "by design" but I meant engineering design. Here's one reference https://github.com/vacp2p/rfc-index/blob/main/status/62/payloads.md/#clock-vs-timestamp-and-message-ordering

It would be better if we double-checked what happens in the desktop app. Maybe they do something different to minimize this problem, but I believe the problem can never be fully eliminated without much deeper changes. @Samyoul please correct me if I'm saying something wrong.

**Comment by ilmotta** [comment]
@osmaczko @jrainville could you shed some light on this discussion? I'm looking for ways to improve this if possible, but messages arriving out of order is a key trait of the messenger, but perhaps Desktop does special handling to mitigate the problem? Thank you

If there's not much to do, we will have to close this issue @mariia-skrypnyk.

**Comment by ilmotta** [comment]
@qfrank do you think there's something substantial we can do to resolve or mitigate this GH issue? Or is it a problem that we shouldn't even attempt to resolve since messages arriving out of order is a feature so to speak?

**Comment by osmaczko** [comment]
@ilmotta sorry for late response..

> [@osmaczko](https://github.com/osmaczko) [@jrainville](https://github.com/jrainville) could you shed some light on this discussion? I'm looking for ways to improve this if possible, but messages arriving out of order is a key trait of the messenger, but perhaps Desktop does special handling to mitigate the problem? Thank you
> 
> If there's not much to do, we will have to close this issue [@mariia-skrypnyk](https://github.com/mariia-skrypnyk).

In this scenario, messages should be fetched from the store node in a single batch, and in principle, we should be able to process them in order. From what I can tell, Waku [delivers batched messages one by one](https://github.com/status-im/status-go/blob/fe28e9c86c4a40e4051a26f16d9a6fd55f174f94/vendor/github.com/waku-org/go-waku/waku/v2/api/history/history.go#L323-L329) to status-go, where they're [queued in filters](https://github.com/status-im/status-go/blob/fe28e9c86c4a40e4051a26f16d9a6fd55f174f94/wakuv2/common/filter.go#L276). These filters are [polled by the messenger once per second](https://github.com/status-im/status-go/blob/fe28e9c86c4a40e4051a26f16d9a6fd55f174f94/services/ext/service.go#L187C14-L187C23), at which point messages are processed and surfaced to the clients.

If the entire batch is pushed from Waku to status-go within the same one-second interval, then we should be able to sort the messages by clock either before handing them to the clients or directly on the client side. If the batch isn’t fully delivered within that timeframe, then there's little we can do without a larger rework. I believe we're facing both issues, though the latter is likely far less impactful.

To sum up: we should sort messages before surfacing them to the UI, either in status-go or on the client. Currently, status-go signals clients with new messages in a [random order](https://github.com/status-im/status-go/blob/fe28e9c86c4a40e4051a26f16d9a6fd55f174f94/protocol/messenger_response.go#L729-L736). There may still be edge cases where even sorting doesn't fully help due to the `waku -> filters -> messenger` machinery, but I expect these to be barely noticeable from the user's perspective.

**Comment by qfrank** [comment]
> [@qfrank](https://github.com/qfrank) do you think there's something substantial we can do to resolve or mitigate this GH issue? Or is it a problem that we shouldn't even attempt to resolve since messages arriving out of order is a feature so to speak?

we can keep this issue open and look into it when we got time 🙏 @ilmotta

**Comment by ilmotta** [comment]
Thanks so much @osmaczko for the explanation. I understand how busy you are with other top priorities, similar to @qfrank. Good thing we can theoretically mitigate this issue.

> Currently, status-go signals clients with new messages in a [random order](https://github.com/status-im/status-go/blob/fe28e9c86c4a40e4051a26f16d9a6fd55f174f94/protocol/messenger_response.go#L729-L736).

About the map keys being in random order, any chance we could use a sorted map data structure directly in status-go to reap the benefits of fast access and predictable iteration?

> If the entire batch is pushed from Waku to status-go within the same one-second interval, then we should be able to sort the messages by clock either before handing them to the clients or directly on the client side. If the batch isn’t fully delivered within that timeframe, then there's little we can do without a larger rework.

I'm sure it was discussed in the past, but I'm wondering what would be the side-effects of increasing the window of time to batch more messages in one go. Would it help with the unordered messages?
